In [1]:
import pandas as pd
import numpy as np
import sklearn
print(sklearn.__version__)

0.20.3


In [2]:
print('S')
print('\n')
print('S')

S


S


In [3]:
location = r'E:\MYLEARN\2-ANALYTICS-DataScience\datasets\employees.csv'

In [4]:
data  = pd.read_csv(location)
data .head()

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0    Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
1     Thomas    Male  3/31/1996         6:53 AM   61933    4.170   
2      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3      Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4      Larry    Male  1/24/1998         4:47 PM  101004    1.389   

  Senior Management             Team  
0              True        Marketing  
1              True              NaN  
2             False          Finance  
3              True          Finance  
4              True  Client Services

In [5]:
data.shape

(1000, 8)

** DataFrame.duplicated() **
In Python’s Pandas library, Dataframe class provides a member function to find duplicate rows based on all columns or some specific columns i.e.

It returns a Boolean Series with True value for each duplicated row

In [6]:
# sorting by first name 
data.sort_values("First Name", inplace = True) 

# making a bool series 
bool_series = data["First Name"].duplicated() 
  
# display data 
data[bool_series] 

First Name  Gender  Start Date Last Login Time  Salary  Bonus %  \
327      Aaron    Male   1/29/1994         6:48 PM   58755    5.097   
440      Aaron    Male   7/22/1990         2:53 PM   52119   11.343   
937      Aaron     NaN   1/22/1986         7:39 PM   63126   18.424   
141       Adam    Male  12/24/1990         8:57 PM  110194   14.727   
302       Adam    Male    7/5/2007        11:59 AM   71276    5.027   
538       Adam    Male   10/8/2010         9:53 PM   45181    3.491   
53        Alan     NaN    3/3/2014         1:28 PM   40341   17.578   
610       Alan    Male   2/17/2012        12:26 AM   41453   10.084   
458     Albert    Male   9/30/2007         5:34 PM  102626   15.843   
477     Albert    Male   8/14/2008         5:43 PM  137840    9.705   
680     Albert    Male   6/17/1992         3:25 AM   86818   14.301   
999     Albert    Male   5/15/2012         6:24 PM  129949   10.169   
959     Albert    Male   9/19/1992         2:35 AM   45094    5.850   
970      Alice  Female    9/3/1988         8:54 PM   63571   15.397   
765      Alice     NaN   4/23/1995         6:35 AM  148339   11.479   
425      Alice  Female    5/2/1986         1:50 AM   51395    2.378   
905      Alice  Female    2/9/2013        12:49 AM  121250    4.363   
684      Alice  Female   1/21/2016         5:07 PM  117787   10.485   
935      Alice  Female   3/23/1996         3:09 AM  131952   12.090   
693      Alice  Female  10/16/1995         9:19 PM   92799    2.782   
642     Amanda     NaN   3/16/2010         9:17 AM  135118   15.108   
578     Amanda  Female   3/17/1982        11:46 PM  107111    1.438   
553     Amanda  Female    6/7/2002         7:46 AM  109290   13.853   
542     Amanda  Female    8/1/2004         1:32 PM   80803   14.077   
641     Amanda  Female   1/25/1995         5:06 AM  102081    5.203   
480        Amy  Female   3/30/2008        10:54 PM  106249   15.967   
467        Amy  Female   6/19/2002         3:06 AM  122897    8.222   
514        Amy     NaN  12/25/1984        10:52 AM   63888   18.115   
699        Amy     NaN   5/19/1984        11:47 AM  102839   10.385   
355     Andrea  Female   9/14/2003        10:30 PM  115913   12.121   
..         ...     ...         ...             ...     ...      ...   
567        NaN  Female    4/1/1980         8:04 PM   48141   12.605   
574        NaN  Female   1/13/2009         1:01 PM  118736    7.421   
599        NaN  Female  10/11/1990        10:57 PM   98385   10.925   
607        NaN    Male  10/13/1983        11:59 PM  139754   12.740   
608        NaN  Female  10/24/1993         5:17 PM  116236   17.274   
626        NaN  Female   4/13/1997         8:03 AM  131755    2.930   
628        NaN     NaN    9/2/1988        12:49 PM  147309    1.702   
674        NaN    Male   8/28/2012         7:56 PM   88733    1.932   
685        NaN    Male  10/20/1989         9:37 AM   88086   19.387   
687        NaN    Male  12/27/1996         7:07 AM  139959    8.992   
697        NaN    Male   5/23/1990         2:09 AM  136655    9.801   
709        NaN    Male    4/7/1995         4:32 AM   60411   12.048   
733        NaN    Male   6/17/2009         1:00 PM  136681    3.655   
734        NaN  Female  12/18/1994         8:42 PM  138807    1.400   
773        NaN    Male  10/24/1986         9:23 AM   47176   10.736   
774        NaN  Female   6/18/2000         7:36 AM  106428   10.867   
782        NaN    Male    5/6/1986         4:41 PM   74104   17.680   
783        NaN     NaN   4/15/1991         3:39 AM  132505   13.592   
785        NaN  Female   7/24/1982        12:00 AM   57811    8.941   
812        NaN    Male  12/13/1994        10:34 AM  141311    5.478   
821        NaN  Female  12/27/1980         2:27 AM   80399   12.254   
823        NaN    Male  11/12/2009         2:30 PM  109411    9.494   
826        NaN     NaN    8/1/1988         1:35 AM   87103    5.665   
877        NaN    Male    9/5/1980         7:36 AM  114896   13.823   
890        NaN    Male  11/2

In [6]:
bool_series

101    False
327     True
440     True
937     True
137    False
141     True
302     True
538     True
300    False
53      True
610     True
372    False
458     True
477     True
680     True
999     True
959     True
988    False
970     True
765     True
425     True
905     True
684     True
935     True
693     True
639    False
642     True
578     True
553     True
542     True
       ...  
567     True
574     True
599     True
607     True
608     True
626     True
628     True
674     True
685     True
687     True
697     True
709     True
733     True
734     True
773     True
774     True
782     True
783     True
785     True
812     True
821     True
823     True
826     True
877     True
890     True
902     True
925     True
946     True
947     True
951     True
Name: First Name, Length: 1000, dtype: bool

In [8]:
#Create a DataFrame
d = {
    'Name':['Alisa','Bobby','jodha','jack','raghu','Cathrine',
            'Alisa','Bobby','kumar','Alisa','Alex','Cathrine'],
    'Age':[26,24,23,22,23,24,26,24,22,23,24,24],
      
    'Score':[85,63,55,74,31,77,85,63,42,62,89,77]}
 
df = pd.DataFrame(data=d, columns=['Name','Age','Score'])
df

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
6      Alisa   26     85
7      Bobby   24     63
8      kumar   22     42
9      Alisa   23     62
10      Alex   24     89
11  Cathrine   24     77

# Duplicates

#### Find the duplicate row in pandas:
duplicated() function is used for find the duplicate rows of the dataframe in python pandas

In [9]:
df.duplicated()
 

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7      True
8     False
9     False
10    False
11     True
dtype: bool

The above code finds whether the row is duplicate and tags TRUE if it is duplicate and tags FALSE if it is not duplicate. 

And assigns it to the column named “is_duplicate”  of the dataframe  df.

In [10]:
df.duplicated('Name')


0     False
1     False
2     False
3     False
4     False
5     False
6      True
7      True
8     False
9      True
10    False
11     True
dtype: bool

This checks if there are duplicate values in a particular column of your DataFrame.

#### Getting rid of duplicates

In [10]:
df.drop_duplicates()

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
8      kumar   22     42
9      Alisa   23     62
10      Alex   24     89

In [12]:
# Sometimes, you may want to drop duplicates just from one column.

df.drop_duplicates(['Name'])

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
8      kumar   22     42
10      Alex   24     89

# checking basic sanity of data

In [12]:
from sklearn import datasets

In [13]:
iris = datasets.load_iris()

In [14]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [15]:
iris.data[:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

In [16]:
print(iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [17]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [18]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [20]:
X = iris.data
y = iris.target

In [21]:
X[:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

In [22]:
y[:5]

array([0, 0, 0, 0, 0])

In [23]:
np.c_[X[:5], y[:5]]

array([[5.1, 3.5, 1.4, 0.2, 0. ],
       [4.9, 3. , 1.4, 0.2, 0. ],
       [4.7, 3.2, 1.3, 0.2, 0. ],
       [4.6, 3.1, 1.5, 0.2, 0. ],
       [5. , 3.6, 1.4, 0.2, 0. ]])

In [24]:
np.c_[np.array([1,2,3]), np.array([4,5,6])]

array([[1, 4],
       [2, 5],
       [3, 6]])

In [25]:
data    = np.c_[X, y]
columns = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'class']

df      = pd.DataFrame(data=data, columns=columns)
df.head()

sepal_length  sepal_width  petal_length  petal_width  class
0           5.1          3.5           1.4          0.2    0.0
1           4.9          3.0           1.4          0.2    0.0
2           4.7          3.2           1.3          0.2    0.0
3           4.6          3.1           1.5          0.2    0.0
4           5.0          3.6           1.4          0.2    0.0

In [26]:
# change the class from numeric to string for easy understanding
replacements  = {0: 'setosa', 1: 'versicolor', 2: 'virginica'}

df['class']   = df['class'].map(replacements)

In [27]:
df.head()

sepal_length  sepal_width  petal_length  petal_width   class
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [28]:
# find out how many samples from each class
df['class'].value_counts()

virginica     50
versicolor    50
setosa        50
Name: class, dtype: int64

In [29]:
# basic statistics
# check for min, max, quartiles, IQR
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [30]:
df[['sepal_length', 'petal_length']].describe()

sepal_length  petal_length
count    150.000000    150.000000
mean       5.843333      3.758000
std        0.828066      1.765298
min        4.300000      1.000000
25%        5.100000      1.600000
50%        5.800000      4.350000
75%        6.400000      5.100000
max        7.900000      6.900000

In [47]:
# calculate the mean petal length of each of the three classes
# df.loc[df['class'] == 'setosa', metrics].mean(axis=0)

In [48]:
# returns indexes
# df['class'] == 'setosa'

In [31]:
# locate by index position
df.loc[[0, 56]]

sepal_length  sepal_width  petal_length  petal_width       class
0            5.1          3.5           1.4          0.2      setosa
56           6.3          3.3           4.7          1.6  versicolor

In [32]:
df.loc[df['class'] == 'setosa']

sepal_length  sepal_width  petal_length  petal_width   class
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8          4.0           1.2          0.2  setosa
15           5.7          4.4           1.5          0.4  setosa
16           5.4          3.9           1.3          0.4  setosa
17           5.1          3.5           1.4          0.3  setosa
18           5.7          3.8           1.7          0.3  setosa
19           5.1          3.8           1.5          0.3  setosa
20           5.4          3.4           1.7          0.2  setosa
21           5.1          3.7           1.5          0.4  setosa
22           4.6          3.6           1.0          0.2  setosa
23           5.1          3.3           1.7          0.5  setosa
24           4.8          3.4           1.9          0.2  setosa
25           5.0          3.0           1.6          0.2  setosa
26           5.0          3.4           1.6          0.4  setosa
27           5.2          3.5           1.5          0.2  setosa
28           5.2          3.4           1.4          0.2  setosa
29           4.7          3.2           1.6          0.2  setosa
30           4.8          3.1           1.6          0.2  setosa
31           5.4          3.4           1.5          0.4  setosa
32           5.2          4.1           1.5          0.1  setosa
33           5.5          4.2           1.4          0.2  setosa
34           4.9          3.1           1.5          0.2  setosa
35           5.0          3.2           1.2          0.2  setosa
36           5.5          3.5           1.3          0.2  setosa
37           4.9          3.6           1.4          0.1  setosa
38           4.4          3.0           1.3          0.2  setosa
39           5.1          3.4           1.5          0.2  setosa
40           5.0          3.5           1.3          0.3  setosa
41           4.5          2.3           1.3          0.3  setosa
42           4.4          3.2           1.3          0.2  setosa
43           5.0          3.5           1.6          0.6  setosa
44           5.1          3.8           1.9          0.4  setosa
45           4.8          3.0           1.4          0.3  setosa
46           5.1          3.8           1.6          0.2  setosa
47           4.6          3.2           1.4          0.2  setosa
48           5.3          3.7           1.5          0.2  setosa
49           5.0          3.3           1.4          0.2  setosa

In [33]:
# display selected columns for class = 'setosa'
df.loc[df['class'] == 'setosa', ['petal_width', 'petal_length']]

petal_width  petal_length
0           0.2           1.4
1           0.2           1.4
2           0.2           1.3
3           0.2           1.5
4           0.2           1.4
5           0.4           1.7
6           0.3           1.4
7           0.2           1.5
8           0.2           1.4
9           0.1           1.5
10          0.2           1.5
11          0.2           1.6
12          0.1           1.4
13          0.1           1.1
14          0.2           1.2
15          0.4           1.5
16          0.4           1.3
17          0.3           1.4
18          0.3           1.7
19          0.3           1.5
20          0.2           1.7
21          0.4           1.5
22          0.2           1.0
23          0.5           1.7
24          0.2           1.9
25          0.2           1.6
26          0.4           1.6
27          0.2           1.5
28          0.2           1.4
29          0.2           1.6
30          0.2           1.6
31          0.4           1.5
32          0.1           1.5
33          0.2           1.4
34          0.2           1.5
35          0.2           1.2
36          0.2           1.3
37          0.1           1.4
38          0.2           1.3
39          0.2           1.5
40          0.3           1.3
41          0.3           1.3
42          0.2           1.3
43          0.6           1.6
44          0.4           1.9
45          0.3           1.4
46          0.2           1.6
47          0.2           1.4
48          0.2           1.5
49          0.2           1.4

In [34]:
# display mean values of selected columns for class = 'setosa'
df.loc[df['class'] == 'setosa', ['petal_width', 'petal_length']].mean()

petal_width     0.246
petal_length    1.462
dtype: float64

In [35]:
# display mean values of selected columns for class = 'setosa'
df.loc[df['class'] == 'setosa', ['petal_width', 'petal_length']].max()

petal_width     0.6
petal_length    1.9
dtype: float64

In [36]:
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [37]:
# use of grouping
groups = df.groupby('class')

In [38]:
# use sum(), min(), max(), count()
groups.mean()

sepal_length  sepal_width  petal_length  petal_width
class                                                           
setosa             5.006        3.428         1.462        0.246
versicolor         5.936        2.770         4.260        1.326
virginica          6.588        2.974         5.552        2.026

In [39]:
groups.agg(['mean', 'min', 'max', 'count'])

sepal_length                 sepal_width                  \
                   mean  min  max count        mean  min  max count   
class                                                                 
setosa            5.006  4.3  5.8    50       3.428  2.3  4.4    50   
versicolor        5.936  4.9  7.0    50       2.770  2.0  3.4    50   
virginica         6.588  4.9  7.9    50       2.974  2.2  3.8    50   

           petal_length                 petal_width                  
                   mean  min  max count        mean  min  max count  
class                                                                
setosa            1.462  1.0  1.9    50       0.246  0.1  0.6    50  
versicolor        4.260  3.0  5.1    50       1.326  1.0  1.8    50  
virginica         5.552  4.5  6.9    50       2.026  1.4  2.5    50

# numeric col containing non-numeric data

In [40]:
d = {
 'unit': ['UD', 'UD', 'UD', 'UD', 'ud','uD'],
 'N-D':  [ 'Q1', 'Q2', 'Q3', 'Q4','Q5','Q6'],
 'num' : [ -1.48, 1.7, -6.18, 0.25, 'weird', 0.25]

}

df = pd.DataFrame(d)
df

unit N-D    num
0   UD  Q1  -1.48
1   UD  Q2    1.7
2   UD  Q3  -6.18
3   UD  Q4   0.25
4   ud  Q5  weird
5   uD  Q6   0.25

In [41]:
df['num'].str.isnumeric()

0      NaN
1      NaN
2      NaN
3      NaN
4    False
5      NaN
Name: num, dtype: object

In [42]:
df['num'].str.isdecimal()

0      NaN
1      NaN
2      NaN
3      NaN
4    False
5      NaN
Name: num, dtype: object

# check for character cases

In [43]:
df['unit'].str.isupper()

0     True
1     True
2     True
3     True
4    False
5    False
Name: unit, dtype: bool

In [44]:
df['unit'] = df['unit'].str.upper()

In [45]:
df

N-D    num unit
0  Q1  -1.48   UD
1  Q2    1.7   UD
2  Q3  -6.18   UD
3  Q4   0.25   UD
4  Q5  weird   UD
5  Q6   0.25   UD

# handling missing values

In [45]:
location = r"E:\MYLEARN\2-ANALYTICS-DataScience\datasets\nba.csv"

In [48]:
# making data frame from csv file 
nba = pd.read_csv(location) 

In [49]:
nba.shape

(457, 9)

In [50]:
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [52]:
# are there nulls?
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary      11
dtype: int64

example #1 - Replacing NaN values with a Static value.

In [24]:
# replacing na values in college with No college 
nba["College"].fillna("No College", inplace = True) 
nba.isnull().sum()
#nba['College'].isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College      0
Salary      11
dtype: int64

Example #2: Using method Parameter

In [25]:
# making data frame from csv file 
nba = pd.read_csv(location) 
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [26]:
# making data frame from csv file 
nba = pd.read_csv(location) 
  
# replacing na values in college with No college 
nba["College"].fillna( method ='ffill', inplace = True) 
  
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4      Georgia State   5000000.0  
5      Georgia State  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

Similarly, bfill, backfill and pad methods can also be used.

Example #3: Using mean value

In [27]:
# making data frame from csv file 
nba = pd.read_csv(location) 

In [28]:
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary      11
dtype: int64

In [29]:
# Or, depending on your needs:
nba['Salary'] = nba['Salary'].fillna(nba['Salary'].mean())
nba['Salary'] = nba['Salary'].fillna(nba['Salary'].min())
nba['Salary'] = nba['Salary'].fillna(nba['Salary'].max())

In [30]:
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary       0
dtype: int64

In [31]:
nba['Age'] = nba['Age'].fillna(200)

In [32]:
nba[nba['Age']==200]

Empty DataFrame
Columns: [Name, Team, Number, Position, Age, Height, Weight, College, Salary]
Index: []

In [33]:
raw_data = {'first_name': ['Jason', np.nan, 'Tina', 'Jake', 'Amy'], 
            'last_name': ['Miller', np.nan, 'Ali', 'Milner', 'Cooze'], 
            'age': [42, np.nan, 36, 24, 73], 
            'sex': ['m', np.nan, 'f', 'm', 'f'], 
            'preTestScore': [4, np.nan, np.nan, 2, 3],
            'postTestScore': [25, np.nan, np.nan, 62, 70]}

df = pd.DataFrame(raw_data, columns = ['first_name', 'last_name', 'age', 'sex', 'preTestScore', 'postTestScore'])
df

first_name last_name   age  sex  preTestScore  postTestScore
0      Jason    Miller  42.0    m           4.0           25.0
1        NaN       NaN   NaN  NaN           NaN            NaN
2       Tina       Ali  36.0    f           NaN            NaN
3       Jake    Milner  24.0    m           2.0           62.0
4        Amy     Cooze  73.0    f           3.0           70.0

In [34]:
# Drop missing observations
# rows returned after dropping, the original df remains intact
df_no_missing = df.dropna()
df_no_missing

first_name last_name   age sex  preTestScore  postTestScore
0      Jason    Miller  42.0   m           4.0           25.0
3       Jake    Milner  24.0   m           2.0           62.0
4        Amy     Cooze  73.0   f           3.0           70.0

In [61]:
df

first_name last_name   age  sex  preTestScore  postTestScore
0      Jason    Miller  42.0    m           4.0           25.0
1        NaN       NaN   NaN  NaN           NaN            NaN
2       Tina       Ali  36.0    f           NaN            NaN
3       Jake    Milner  24.0    m           2.0           62.0
4        Amy     Cooze  73.0    f           3.0           70.0

In [62]:
#Drop rows where all cells in that row is NA
df_cleaned = df.dropna(how='all')
df_cleaned

first_name last_name   age sex  preTestScore  postTestScore
0      Jason    Miller  42.0   m           4.0           25.0
2       Tina       Ali  36.0   f           NaN            NaN
3       Jake    Milner  24.0   m           2.0           62.0
4        Amy     Cooze  73.0   f           3.0           70.0

In [64]:
# Create a new column full of missing values
df['location'] = np.nan
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           NaN            NaN       NaN
2       Tina       Ali  36.0    f           NaN            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [66]:
# Drop column if they only contain missing values
df.dropna(axis=1, how='all')

first_name last_name   age  sex  preTestScore  postTestScore
0      Jason    Miller  42.0    m           4.0           25.0
1        NaN       NaN   NaN  NaN           NaN            NaN
2       Tina       Ali  36.0    f           NaN            NaN
3       Jake    Milner  24.0    m           2.0           62.0
4        Amy     Cooze  73.0    f           3.0           70.0

In [67]:
# Drop rows that contain less than five observations
# This is really mostly useful for time series
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           NaN            NaN       NaN
2       Tina       Ali  36.0    f           NaN            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [107]:
df.dropna(thresh=5)

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       NaN
3       Jake    Milner  24.0   m           2.0           62.0       NaN
4        Amy     Cooze  73.0   f           3.0           70.0       NaN

In [90]:
# Fill in missing data with zeros
df.fillna(0)

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       0.0
1          0         0   0.0   0           0.0            0.0       0.0
2       Tina       Ali  36.0   f           0.0            0.0       0.0
3       Jake    Milner  24.0   m           2.0           62.0       0.0
4        Amy     Cooze  73.0   f           3.0           70.0       0.0

In [112]:
# Fill in missing in preTestScore with the mean value of preTestScore
# inplace=True means that the changes are saved to the df right away

In [91]:
df["preTestScore"].fillna(df["preTestScore"].mean(), inplace=True)
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           3.0            NaN       NaN
2       Tina       Ali  36.0    f           3.0            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [115]:
# Fill in missing in postTestScore with each sex’s mean value of postTestScore
df["postTestScore"].fillna(df.groupby("sex")["postTestScore"].transform("mean"), inplace=True)
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           3.0            NaN       NaN
2       Tina       Ali  36.0    f           3.0           70.0       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

#### using scikit learn imputer

strategy : string, optional (default=”mean”)
The imputation strategy.

- If “mean”, then replace missing values using the mean along each column. Can only be used with numeric data.
- If “median”, then replace missing values using the median along each column. Can only be used with numeric data.
- If “most_frequent”, then replace missing using the most frequent value along each column. Can be used with strings or numeric data.
- If “constant”, then replace missing values with fill_value. Can be used with strings or numeric data.

In [108]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import Imputer

df = pd.DataFrame([[10,     20],
                   [np.nan, 23],
                   [15,    np.nan],
                   [17,    29]])

df.dtypes

0    float64
1    float64
dtype: object

In [109]:
imp = SimpleImputer(strategy="median")

print(imp.fit_transform(df)) 

[[10. 20.]
 [15. 23.]
 [15. 23.]
 [17. 29.]]


one more example ...

In [110]:
X = np.array([[23.56],
              [53.45],
              ['NaN'],
              [44.44],
              [77.78],
              ['NaN'],
              [234.44],
              [11.33],
              [79.87]])

In [111]:
imputer = SimpleImputer(strategy='mean')  

imputer.fit_transform(X)

array([[ 23.56      ],
       [ 53.45      ],
       [ 74.98142857],
       [ 44.44      ],
       [ 77.78      ],
       [ 74.98142857],
       [234.44      ],
       [ 11.33      ],
       [ 79.87      ]])

one more example ..

In [112]:
# Create an empty dataset
df = pd.DataFrame()

# Create two variables called x0 and x1. Make the first value of x1 a missing value
df['x0'] = [0.3051,0.4949,0.6974,0.3769,0.2231,0.341,0.4436,0.5897,0.6308,0.5]
df['x1'] = [np.nan,0.2654,0.2615,0.5846,0.4615,0.8308,0.4962,0.3269,0.5346,0.6731]

# View the dataset
df

x0      x1
0  0.3051     NaN
1  0.4949  0.2654
2  0.6974  0.2615
3  0.3769  0.5846
4  0.2231  0.4615
5  0.3410  0.8308
6  0.4436  0.4962
7  0.5897  0.3269
8  0.6308  0.5346
9  0.5000  0.6731

In [113]:
# Fit Imputer
# Create an imputer object that looks for 'Nan' values, then replaces them with the mean value of the feature by columns (axis=0)
mean_imputer = Imputer(missing_values='NaN', strategy='mean', axis=0)

# Train the imputor on the df dataset
mean_imputer = mean_imputer.fit(df)

C:\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:58: DeprecationWarning: Class Imputer is deprecated; Imputer was deprecated in version 0.20 and will be removed in 0.22. Import impute.SimpleImputer from sklearn instead.
  warnings.warn(msg, category=DeprecationWarning)


In [114]:
# Apply Imputer
# Apply the imputer to the df dataset
imputed_df = mean_imputer.transform(df.values)

In [115]:
imputed_df

array([[0.3051    , 0.49273333],
       [0.4949    , 0.2654    ],
       [0.6974    , 0.2615    ],
       [0.3769    , 0.5846    ],
       [0.2231    , 0.4615    ],
       [0.341     , 0.8308    ],
       [0.4436    , 0.4962    ],
       [0.5897    , 0.3269    ],
       [0.6308    , 0.5346    ],
       [0.5       , 0.6731    ]])

Notice that 0.49273333 is the imputed value, replacing the np.NaN value.

#### more examples ... multiple columns

In [116]:
titanic = pd.read_csv("https://raw.githubusercontent.com/amueller/scipy-2017-sklearn/master/notebooks/datasets/titanic3.csv")
# there is still a small problem with using the OneHotEncoder and missing values,
# so for now I am going to assume there are no missing values by dropping them

titanic2 = titanic.dropna(subset=['pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked'])

In [117]:
target   = titanic2.survived.values
features = titanic2[['pclass', 'sex', 'age', 'fare', 'embarked']]

In [118]:
features.head()

pclass     sex      age      fare embarked
0       1  female  29.0000  211.3375        S
1       1    male   0.9167  151.5500        S
2       1  female   2.0000  151.5500        S
3       1    male  30.0000  151.5500        S
4       1  female  25.0000  151.5500        S

In [119]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer, make_column_transformer

In [120]:
preprocess = make_column_transformer(
    (StandardScaler(), ['age', 'fare']),
    (OneHotEncoder(),  ['pclass', 'sex', 'embarked'])
)

In [121]:
new_features = preprocess.fit_transform(features)
new_features[:3]

array([[-0.05663194,  3.13554913,  1.        ,  0.        ,  0.        ,
         1.        ,  0.        ,  0.        ,  0.        ,  1.        ],
       [-2.01237899,  2.06268333,  1.        ,  0.        ,  0.        ,
         0.        ,  1.        ,  0.        ,  0.        ,  1.        ],
       [-1.93693697,  2.06268333,  1.        ,  0.        ,  0.        ,
         1.        ,  0.        ,  0.        ,  0.        ,  1.        ]])

In [122]:
new_features_df = pd.DataFrame(data=new_features)
new_features_df.head()

0         1    2    3    4    5    6    7    8    9
0 -0.056632  3.135549  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0
1 -2.012379  2.062683  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0
2 -1.936937  2.062683  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0
3  0.013009  2.062683  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0
4 -0.335196  2.062683  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0